crypto/stark-verifier-ethereum/contracts/fri.sol

```diff
@@ -95,25 +109,40 @@ contract Fri is Trace, MerkleVerifier {
     // Note the final layer folded values will be overwritten to the input data locations.
     function fold_and_check_fri_layers(FriContext memory fri_data) internal {
         trace('fold_and_check_fri_layers', true);
-        PrimeField.EvalX memory eval = PrimeField.init_eval(fri_data.log_eval_domain_size);
         LayerContext memory layer_context = LayerContext({
             len: uint64(2)**(fri_data.log_eval_domain_size),
             step: 1,
-            coset_size: 0
+            coset_size: uint64(2)**(fri_data.fri_layout[0]),
+            generator: PrimeField.generator_power(fri_data.log_eval_domain_size),
+            log_domain_size: fri_data.log_eval_domain_size,
+            x_inv: new uint256[](fri_data.queries.length),
+            roots: [1, OROOT4, OROOT2, OROOT6, OROOT1, OROOT5, OROOT3, OROOT7]
         });
         uint256[] memory merkle_indices = new uint256[](fri_data.queries.length);
         bytes32[] memory merkle_val = new bytes32[](fri_data.queries.length);
 
+        // Initialize x_inv
+        trace('init_x_inv', true);
+        for (uint256 i = 0; i < fri_data.queries.length; i++) {
+            uint256 index = fri_data.queries[i];
+            index = index.bit_reverse2(layer_context.log_domain_size);
+            index = layer_context.len - index;
+            layer_context.x_inv[i] = layer_context.generator.fpow(index);
+        }
+        trace('init_x_inv', false);
+
+        // Fold layers
         for (uint256 i = 0; i < fri_data.fri_layout.length; i++) {
             layer_context.coset_size = uint64(2)**(fri_data.fri_layout[i]);
+            require(layer_context.coset_size <= MAX_COSET_SIZE, 'Coset too large');
+
             // Overwrites and resizes the data array and the querry index array
             // They will contain the folded points and indexes
             fold_layer(
                 fri_data.polynomial_at_queries,
                 fri_data.queries,
-                Iterators.init_iterator(fri_data.fri_values[i]),
-                eval,
-                fri_data.eval_points[i],
+                fri_data.fri_values[i],
+                fri_data.eval_points[i].from_montgomery(),
                 layer_context,
                 merkle_val
             );
```

Comment:
The deep_copy_and_convert function was named that because it went from uint64 -> uint256, we should change the name of the function to 'deep_copy' now that the uint64 has been factored out.

Reply:
`merkle_indices` is `uint256[]` and `fri_data.queries` is `uint64[]` so conversion still occurs.